### PR TITLE
use a more 'normal' storage factory

### PR DESCRIPTION
**Pull request by deads2k**
This attempts to remove our special storage factory and collapse onto the same one that upstream uses.  It currently fails tests and needs at least one more, so WIP.

@enj this is why I'm worried about that CR test.

/assign @sttts

### CODE REVIEW

File: pkg/cmd/openshift-apiserver/openshiftapiserver/configprocessing/etcd.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import (
+	configapi "github.com/openshift/origin/pkg/cmd/server/apis/config"
+	cmdflags "github.com/openshift/origin/pkg/cmd/util/flags"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apiserver/pkg/server/options"
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+)
+
+func GetEtcdOptions(startingFlags map[string][]string, etcdConnectionInfo configapi.EtcdConnectionInfo, storagePrefix string, defaultWatchCacheSizes map[schema.GroupResource]int) (*options.EtcdOptions, error) {
```
Comment (sttts) [inline pkg/cmd/openshift-apiserver/openshiftapiserver/configprocessing/etcd.go]: godoc. startignFlags are api server flags?

File: pkg/cmd/openshift-apiserver/openshiftapiserver/configprocessing/etcd.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	storageConfig.Type = "etcd3"
+	storageConfig.Quorum = true
+	storageConfig.Prefix = storagePrefix
+	storageConfig.ServerList = etcdConnectionInfo.URLs
+	storageConfig.KeyFile = etcdConnectionInfo.ClientCert.KeyFile
+	storageConfig.CertFile = etcdConnectionInfo.ClientCert.CertFile
+	storageConfig.CAFile = etcdConnectionInfo.CA
+
+	etcdOptions := options.NewEtcdOptions(storageConfig)
+	etcdOptions.DefaultStorageMediaType = "application/json" // TODO(post-1.6.1-rebase): enable protobuf with etcd3 as upstream
```
Comment (sttts) [inline pkg/cmd/openshift-apiserver/openshiftapiserver/configprocessing/etcd.go]: no proto?

File: pkg/cmd/server/origin/rest/storage_options.go [hunk 3]
```diff
@@ -1,59 +0,0 @@
...
-			{Resource: "egressnetworkpolicies", Group: "network.openshift.io"}: "registry/egressnetworkpolicy",
-			{Resource: "hostsubnets"}:                                          "registry/sdnsubnets",
-			{Resource: "hostsubnets", Group: "network.openshift.io"}:           "registry/sdnsubnets",
-			{Resource: "netnamespaces"}:                                        "registry/sdnnetnamespaces",
-			{Resource: "netnamespaces", Group: "network.openshift.io"}:         "registry/sdnnetnamespaces",
-		},
-		// storage versions: no overrides anymore
-		map[schema.GroupResource]schema.GroupVersion{},
-		// quorum resources:
-		map[schema.GroupResource]struct{}{
```
Comment (sttts) [inline pkg/cmd/server/origin/rest/storage_options.go]: this special quorum handling is gone, isn't it? So we default to quorum=true?

File: pkg/cmd/util/flags/flags.go [hunk 4]
```diff
@@ -33,7 +35,13 @@ func Apply(args map[string][]string, flags *pflag.FlagSet) []error {
 func Resolve(args map[string][]string, fn func(*pflag.FlagSet)) []error {
 	fs := pflag.NewFlagSet("extended", pflag.ContinueOnError)
 	fn(fs)
-	return Apply(args, fs)
+	return apply(args, fs, false)
+}
+
+func ResolveIgnoreMissing(args map[string][]string, fn func(*pflag.FlagSet)) []error {
```
Comment (sttts) [inline pkg/cmd/util/flags/flags.go]: godoc

**Comment by deads2k** [comment]
/hold

holding to make sure we don't accidentally merge a json backed storage impl.

@enj Any idea why the proto deserialization doesn't work?  I'm looking and I can't see where you generated the proto files for your type, so how did that ever work?  reflective conversion?

**Comment by deads2k** [comment]
/hold cancel

This is ready for review/merge now.

File: pkg/oc/clusterup/coreinstall/kubeapiserver/config.go [hunk 5]
```diff
@@ -80,8 +80,6 @@ func (opt KubeAPIServerStartConfig) MakeMasterConfig(dockerClient dockerhelper.I
 		return "", err
 	}
 
-	masterconfig.KubernetesMasterConfig.APIServerArguments["feature-gates"] = []string{"CustomResourceSubresources=true"}
```
Comment (deads2k) [inline pkg/oc/clusterup/coreinstall/kubeapiserver/config.go]: this is now on by default.

File: pkg/util/restoptions/configgetter.go [hunk 6]
```diff
@@ -1,119 +1,50 @@
...
-	storageFactory, err := kubernetes.BuildStorageFactory(apiserverOptions, enforcedStorageVersions)
-	if err != nil {
-		return nil, err
-	}
-	storageFactory.DefaultResourcePrefixes = resourcePrefixOverrides
-	storageFactory.StorageConfig.Prefix = masterOptions.EtcdStorageConfig.OpenShiftStoragePrefix
-
-	// perform watch cache heuristic like upstream
-	if apiserverOptions.Etcd.EnableWatchCache {
-		glog.V(2).Infof("Initializing cache sizes based on %dMB limit", apiserverOptions.GenericServerRunOptions.TargetRAMMB)
```
Comment (sttts) [inline pkg/util/restoptions/configgetter.go]: the log message is gone. Intentionally?

**Comment by sttts** [comment]
Have tried to follow the path of all options. Looks good so far. Just a question about the log output. Otherwise lgtm.

File: pkg/util/restoptions/configgetter.go (hunk 6 above)
Comment (deads2k) [reply pkg/util/restoptions/configgetter.go]: > the log message is gone. Intentionally?

Yeah.  If we need to expose it, we need to expose it a different way.  Metric probably.

**Pull request closed without merging**